ckan/lib/webassets_tools.py

```diff
@@ -1,172 +1,244 @@
+
     if not os.path.exists(config_path):
+        log.debug(
+            "Cannot create library %s at %s because webassets.yaml is missing",
+            name,
+            path,
+        )
         return
-    assert env
+
     library: dict[str, Any] = YAMLLoader(config_path).load_bundles()
     bundles = {
-        u'/'.join([name, key]): bundle
+        f"{name}/{key}": bundle
         for key, bundle
         in library.items()
     }
 
-    # Unfortunately, you'll get an error attempting to register bundle
-    # with the same name twice. For now, let's just pop existing
-    # bundle and avoid name-conflicts
-    # TODO: make PR into webassets with preferable solution
-    # Issue: https://github.com/miracle2k/webassets/issues/519
+    # skip attempts to register an asset if name is taken. It gives us
+    # templates-like behavior, where the item that was registered first has
+    # highest priority.
     for name, bundle in bundles.items():
-        env._named_bundles.pop(name, None)
+        if is_registered(name):
+            log.debug(
+                "Skip registration of %s because it already exists",
+                name,
+            )
+            continue
+
+        # use absolute path to files. Otherwise they'll behave like templates
+        # and if plugin A and plugin B has their own `x.css`(i.e
+        # `ckanext/A/assets/x.css` and `ckanext/B/assets/x.css`), both plugins
+        # will load the same `x.css` from the plugin that was first registered
+        bundle.contents = [
+            os.path.join(path, item)
+            for item in bundle.contents
+        ]
+        log.debug("Register asset %s", name)
         env.register(name, bundle)
 
-    env.append_path(path)
-
 
 def webassets_init() -> None:
+    """Initialize fresh Webassets environment
+    """
     global env
 
     static_path = get_webassets_path()
 
-    public = config.get(u'ckan.base_public_folder')
-
-    public_folder = os.path.abspath(os.path.join(
-        os.path.dirname(__file__), u'..', public))
-
-    base_path = os.path.join(public_folder, u'base')
-
     env = Environment()
     env.directory = static_path
-    env.debug = config.get(u'debug')
-    env.url = u'/webassets/'
+    env.debug = config["debug"] and False
+    env.url = config["ckan.webassets.url"]
 
-    add_public_path(base_path, u'/base/')
 
-    logger.debug(u'Base path {0}'.format(base_path))
-    create_library(u'vendor', os.path.join(
-        base_path, u'vendor'))
+def register_core_assets():
+    """Register CKAN core assets.
 
-    create_library(u'base', os.path.join(base_path, u'javascript'))
+    Call this function after registration of plugin assets. Asset overrides are
+    not alowed, so if plugin tries to replace CKAN core asset, it has to
+    register an asset with the same name before core asset is added. In this
+    case, asset from plugin will have higher precedence and core asset will be
+    ignored.
```

Comment:
Here it says: `Asset overrides are not allowed` but then explains that if plugins tries to replace CKAN core asset, they can.

Reply:
Yes. Webassets library does not allow overrides. If we want to use our custom asset instead of the existing one, we have to register it earlier. In this way, our asset is registered, and when core asset is added to registry, we ignore it, because overrides are not possible